handlers/oauth/oauth.go

```diff
@@ -0,0 +1,142 @@
+package oauth
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgerrcode"
+	"github.com/markbates/goth"
+	"github.com/markbates/goth/gothic"
+	"github.com/markbates/goth/providers/twitter"
+	"github.com/speedrun-website/leaderboard-backend/model"
+)
+
+var twitterProvider = twitter.NewAuthenticate(
+	os.Getenv("TWITTER_OAUTH_KEY"),
+	os.Getenv("TWITTER_OAUTH_SECRET"),
+	os.Getenv("TWITTER_OAUTH_CALLBACK_URL"),
+)
+
+var completeUserAuth = gothic.CompleteUserAuth
+
+func InitializeProviders() {
+	enabledProviders := strings.Split(os.Getenv("ENABLED_PROVIDERS"), ",")
+	for _, provider := range enabledProviders {
+		cleanedProvider := strings.ToUpper(strings.TrimSpace(provider))
+		switch cleanedProvider {
+		case twitterProvider.Name():
+			log.Println("enabling twitter provider")
+			goth.UseProviders(twitterProvider)
+		default:
+			log.Printf("Unknown provider %s", cleanedProvider)
+		}
+	}
+	goth.UseProviders(twitterProvider)
+}
+
+//TODO Replace with generic error response struct
+type OauthErrorResponse struct {
+	Error string `json:"error"`
+}
+
+//OauthLogin begins the oauth authentication process
+//such as fetching tokens and redirecting the user to the providers website.
+func OauthLogin(c *gin.Context) {
+	log.Printf("%s oauth authentication", c.Query("provider"))
+	// Handles redirecting the user
+	gothic.BeginAuthHandler(c.Writer, c.Request)
+}
+
+//OauthCallback handles the oauth(1.0a/2) callback mechanism
+//The frontend needs to make sure to append ?provider={provider}
+func OauthCallback(c *gin.Context) {
+	//TODO: Connect to JWT
+	provider := c.Query("provider")
+	log.Printf("%s oauth callback", provider)
+
+	providerUserData, err := completeUserAuth(c.Writer, c.Request)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, &OauthErrorResponse{
+			Error: err.Error(),
+		})
+		return
+	}
+
+	var existingUser *model.User
+	var userLookupErr error
+	if provider == twitterProvider.Name() {
+		existingUser, userLookupErr = Oauth.GetUserByTwitterID(providerUserData.UserID)
+	} else {
+		c.AbortWithStatusJSON(http.StatusBadRequest, OauthErrorResponse{
+			Error: fmt.Sprintf("Unsupported provider %s", provider),
+		})
+		return
+	}
+
+	if userLookupErr != nil {
+		log.Printf("error looking up user: %s", userLookupErr)
+		c.AbortWithStatusJSON(http.StatusInternalServerError, OauthErrorResponse{
+			Error: "Issue finding user",
+		})
+		return
+	}
+
+	if existingUser != nil {
+		c.JSON(http.StatusOK,
+			//We "copy" the struct here to ensure that response is consistent
+			//even if the type of the existing user changes
+			&model.UserIdentifier{
```

Comment:
Note: This will be moved to using the `data` top level key added in #60, we can fix it in the PR that will align this work to that PR. So it should be fine not to block this PR on that for now.